moveit_core/trajectory_processing/src/ruckig_traj_smoothing.cpp

```diff
@@ -132,61 +117,142 @@ bool RuckigSmoothing::applySmoothing(robot_trajectory::RobotTrajectory& trajecto
   }
 
   ruckig::Result ruckig_result;
-  for (size_t waypoint_idx = 0; waypoint_idx < num_waypoints - 1; ++waypoint_idx)
+  bool smoothing_complete = false;
+  robot_trajectory::RobotTrajectory original_trajectory_copy(trajectory);
+  size_t duration_extension_attempts = 0;
+  while (rclcpp::ok() && (duration_extension_attempts < MAX_DURATION_EXTENSION_ATTEMPTS) && !smoothing_complete)
   {
-    moveit::core::RobotStatePtr next_waypoint = trajectory.getWayPointPtr(waypoint_idx + 1);
+    for (size_t waypoint_idx = 0; waypoint_idx < num_waypoints - 1; ++waypoint_idx)
+    {
+      moveit::core::RobotStatePtr next_waypoint = trajectory.getWayPointPtr(waypoint_idx + 1);
 
-    getNextCurrentTargetStates(ruckig_input, ruckig_output, next_waypoint, num_dof, idx);
+      getNextCurrentTargetStates(ruckig_input, ruckig_output, next_waypoint, num_dof, idx);
 
-    // Run Ruckig
-    ruckig_result = ruckig.update(ruckig_input, ruckig_output);
+      // Run Ruckig
+      ruckig_result = ruckig.update(ruckig_input, ruckig_output);
 
-    // If the requested velocity is too great, a joint can actually "move backward" to give itself more time to
-    // accelerate to the target velocity. Iterate and decrease velocities until that behavior is gone.
-    bool backward_motion_detected = checkForLaggingMotion(num_dof, ruckig_input, ruckig_output);
+      // If the requested velocity is too great, a joint can actually "move backward" to give itself more time to
+      // accelerate to the target velocity. Iterate and decrease velocities until that behavior is gone.
+      bool backward_motion_detected = checkForLaggingMotion(num_dof, ruckig_input, ruckig_output);
 
-    double minimum_velocity_magnitude = 0.01;  // rad/s
-    double velocity_magnitude = getTargetVelocityMagnitude(ruckig_input, num_dof);
-    while (backward_motion_detected && rclcpp::ok() && (velocity_magnitude > minimum_velocity_magnitude))
-    {
-      // decrease target velocity
+      double minimum_velocity_magnitude = 0.01;  // rad/s
+      double velocity_magnitude = getTargetVelocityMagnitude(ruckig_input, num_dof);
+      while (backward_motion_detected && rclcpp::ok() && (velocity_magnitude > minimum_velocity_magnitude))
+      {
+        // Skip repeated waypoints with no change in position. Ruckig does not handle this well and there's really no
+        // need to smooth it Simply set it equal to the previous (identical) waypoint.
+        if (checkForIdenticalWaypoints(*trajectory.getWayPointPtr(waypoint_idx), *next_waypoint, num_dof, idx))
+        {
+          *next_waypoint = trajectory.getWayPoint(waypoint_idx);
+          continue;
+        }
+
+        // decrease target velocity
+        for (size_t joint = 0; joint < num_dof; ++joint)
+        {
+          ruckig_input.target_velocity.at(joint) *= 0.9;
```

Comment:
I don't really see how these lines could be combined. L153 updates velocity, L155 updates acceleration.

Roger that on using Eigen vectors to avoid the for loop, but... then I would have to initialize an Eigen type for each of these variables. Especially since the vectors are pretty short (~6-7 elements), I don't think that would help performance.